File: showTournamentMatchup.rb
require 'libRuby'
gStyle.SetOptStat(0)
gStyle.SetTitleFillColor(10)
gStyle.SetTitleBorderSize(0)
gStyle.SetPalette(1,0)

def printOptions
  puts ""
  puts "showTournamentMatchup.rb Usage options"
  puts ""
  puts "\t-F (string) currently existing file name, or name of output ROOT file (no default)[Optional]"
  puts "\t-t (string) teamnameA (no default)[Required if -F not used]"
  puts "\t-T (string) teamnameB (no default)[Required if -F not used]"
  puts "\t-H calculate and plot gameScore percentages"
  puts "\t-s (int) (double) srs value to use for weights and histograms (default: \"fixed\")[Optional]"
  puts "\t-h print this message"
  puts ""
  puts "Ex: ruby showTournamentMatchup.rb -t \"2015 north carolina\" -T \"2015 harvard\" -H -s 0.5"
  puts
  puts "This program is used to show the results of the showTournamentMatchup program, but"
  puts "it can also be used to run that program first and then process the output, so as to make"
  puts "this a one-step process.  If a ROOT file already exists, you can plot its graphs"
  puts "using the -F switch to set the input file name. If a ROOT file does not already exist,"
  puts "then you can create one by using the -t and -T options to set the names of the teams you want to plot."
  puts "If you wish to save the ROOT file created, simply use the"
  puts "-F option to set the name, it will be saved by default in the rootFiles/ directory, so"
  puts "only the name is necessary.  If no name is set, then a temporary file is created which"
  puts "is deleted at the end of the program."
  puts ""
end

fileName = nil
teamNameA = nil
teamNameB = nil
srsVal = "free"
useHistograms = false

#command line switcher
ARGV.each_with_index do |entry, index|
  case entry
  when /^-F/
    fileName = ARGV[index+1]
  when /^-s/
    srsVal = ARGV[index+1]
  when /^-H/
    useHistograms = true
  when /^-t/
    teamNameA = ARGV[index+1]
  when /^-T/
    teamNameB = ARGV[index+1]
  end
end

file = nil
if (teamNameA.nil? && teamNameB.nil?)
  if fileName.nil?
    puts "Must enter teamNameA and teamNameB or a fileName"
    exit
  else
    file = TFile.new(fileName)
  end
else
  if fileName.nil?
    if !useHistograms
      `$CLION/showTournamentMatchup -t \"#{teamNameA}\" -T \"#{teamNameB}\" -o \"temp.root\"`
    else
      `$CLION/showTournamentMatchup -t \"#{teamNameA}\" -T \"#{teamNameB}\" -o \"temp.root\" -H -s #{srsVal}`
    end
    #sleep 3
    file = TFile.new("rootFiles/temp.root")
  else
    if !useHistograms
      `$CLION/showTournamentMatchup -t \"#{teamNameA}\" -T \"#{teamNameB}\" -o \"#{fileName}\"`
    else
      `$CLION/showTournamentMatchup -t \"#{teamNameA}\" -T \"#{teamNameB}\" -o \"#{fileName}\" -H -s #{srsVal}`
    end
    #sleep 3
    file = TFile.new("rootFiles/#{fileName}")
  end
end

avgsA_per_time = {}
wavgsA_per_time = {}
avgsB_per_time = {}
wavgsB_per_time = {}
league_avgs = {}
lines = {}
realLines = {}
ss = ["oor.p","oefg.p","oftmr.p","oto.p",
      "dor.p","defg.p","dftmr.p","dto.p","srs","rpi"]

ss.each do |s|
  avgsA_per_time[s] = gROOT.FindObject("avgA_per_time_#{s.gsub(".","")}")
  wavgsA_per_time[s] = gROOT.FindObject("wavgA_per_time_#{s.gsub(".","")}")

  avgsB_per_time[s] = gROOT.FindObject("avgB_per_time_#{s.gsub(".","")}")
  wavgsB_per_time[s] = gROOT.FindObject("wavgB_per_time_#{s.gsub(".","")}")

  #For some idiot reason I don't understand I can't draw a line read in this way
  #so I have to make a new TLine object using it.
  lines[s] = gROOT.FindObject("#{s.gsub(".","")}Line")
  realLines[s] = TLine.new(lines[s].GetX1,lines[s].GetY1,lines[s].GetX2,lines[s].GetY2)
end

if useHistograms
  avgsA_per_time["gameScore"] = gROOT.FindObject("avgA_per_time_gameScore")
  wavgsA_per_time["gameScore"] = gROOT.FindObject("wavgA_per_time_gameScore")

  avgsB_per_time["gameScore"] = gROOT.FindObject("avgB_per_time_gameScore")
  wavgsB_per_time["gameScore"] = gROOT.FindObject("wavgB_per_time_gameScore")

  realLines["gameScore"] = realLines["rpi"]
end

can = TCanvas.new("can","",0,0,1300,700)
can.Divide(4,3)

ss.each_with_index do |s,i|
  can.cd(i+1)
  wavgsA_per_time[s].SetMinimum(0)
  wavgsA_per_time[s].SetMaximum(1)
  wavgsA_per_time[s].SetMinimum(-30) if s == "srs"
  wavgsA_per_time[s].SetMaximum(30) if s == "srs"
  wavgsA_per_time[s].SetMarkerStyle(4)
  wavgsA_per_time[s].SetMarkerColor(4)
  wavgsA_per_time[s].SetLineColor(4)
  wavgsA_per_time[s].SetMarkerSize(0.5)
  wavgsA_per_time[s].Draw("psame")
  # avgsA_per_time[s].SetMarkerStyle(8)
  # avgsA_per_time[s].SetMarkerColor(4)
  # avgsA_per_time[s].SetMarkerSize(0.5)
  # avgsA_per_time[s].Draw("psame")

  wavgsB_per_time[s].SetMarkerStyle(4)
  wavgsB_per_time[s].SetMarkerColor(2)
  wavgsB_per_time[s].SetLineColor(2)
  wavgsB_per_time[s].SetMarkerSize(0.5)
  wavgsB_per_time[s].Draw("psame")

  realLines[s].SetLineColor(2)
  realLines[s].Draw("same")
end

if useHistograms
  can.cd(11)
  wavgsA_per_time["gameScore"].SetMinimum(0)
  wavgsA_per_time["gameScore"].SetMaximum(1)
  wavgsA_per_time["gameScore"].SetMarkerStyle(4)
  wavgsA_per_time["gameScore"].SetMarkerColor(4)
  wavgsA_per_time["gameScore"].SetLineColor(4)
  wavgsA_per_time["gameScore"].SetMarkerSize(0.5)
  wavgsA_per_time["gameScore"].Draw("psame")
  # avgsA_per_time["gameScore"].SetMarkerStyle(8)
  # avgsA_per_time["gameScore"].SetMarkerColor(4)
  # avgsA_per_time["gameScore"].SetMarkerSize(0.5)
  # avgsA_per_time["gameScore"].Draw("psame")

  wavgsB_per_time["gameScore"].SetMarkerStyle(4)
  wavgsB_per_time["gameScore"].SetMarkerColor(2)
  wavgsB_per_time["gameScore"].SetLineColor(2)
  wavgsB_per_time["gameScore"].SetMarkerSize(0.5)
  wavgsB_per_time["gameScore"].Draw("psame")

  realLines["gameScore"].SetLineColor(2)
  realLines["gameScore"].Draw("same")
end

if (!teamNameA.nil? and fileName.nil?)
  `rm rootFiles/temp.root`
end

##This sets it so that the application ends when a canvas is closed. 
##Without it the application can hang in the background if you 
##don't exit properly.
gApplication.Connect("TCanvas","Closed()","TApplication",gApplication,
                     "Terminate()")
##If runROOT selected the window will stay until closed,  
##If not selected the window goes away after the fits run. 
 gApplication.Run# if runROOT             

File: showTournamentMatchup.cpp
#include <iostream>
#include <fstream>
#include "TFile.h"
#include "TH1F.h"
#include "TLine.h"
#include <boost/algorithm/string.hpp>
#include <TROOT.h>
#include "src/Team.h"
#include "src/ConstantTeam5YearAverages.h"
#include "src/ConstantSeasonInfo.h"
#include "src/ConstantGameFunction.h"
#include "src/ConstantSRSadditions.h"
#include "src/ConstantTeamNeutralRatios.h"
#include "src/ConstantWAverageFunctions.h"
#include "src/ConstantStandardDeviations.h"

void printOptions();

int main(int argc,char *argv[]){
    int c;
    int year = 0;
    std::string teamNameA = "", teamNameB = "", outFileName = "";
    bool useHistogramsFile = false;
    std::string srsValue = "free";

    /*____________________________Parse Command Line___________________________*/
    while((c = getopt(argc,argv,"t:T:o:hHs:")) != -1){
        switch(c){
            case 't':
                teamNameA = optarg;
                break;
            case 'T':
                teamNameB = optarg;
                break;
            case 'o':
                outFileName = optarg;
                break;
            case 'H':
                useHistogramsFile = true;
                break;
            case 's':
                srsValue.assign(optarg);
                break;
            case 'h':
                printOptions();
                return 0;
            default:
                // not an option
                break;
        }
    }

    if (teamNameA == "" || teamNameB == "" || outFileName == ""){
        std::cout << "You must specify one team with the -t switch " << std::endl;
        std::cout << "and another team with the -T switch" << std::endl;
        std::cout << "and an outFileName with the -o switch" << std::endl;
        return 0;
    }

    year = stoi(teamNameA.substr(0,4));

    char *homePath, path[256], histName[256];
    homePath = getenv("HOME");

    sprintf(path, "%s/cpp/NCAA_C/constants/team_5yr_averages.d", homePath);
    ConstantTeam5YearAverages *averages = ConstantTeam5YearAverages::Instance();
    averages->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/season_info.d", homePath);
    ConstantSeasonInfo *seasonInfo = ConstantSeasonInfo::Instance();
    seasonInfo->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/waverage_functions.d", homePath);
    ConstantWAverageFunctions *functions = ConstantWAverageFunctions::Instance();
    functions->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/team_neutral_ratios.d", homePath);
    ConstantTeamNeutralRatios *ratios = ConstantTeamNeutralRatios::Instance();
    ratios->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/srs_additions.d", homePath);
    ConstantSRSadditions *additions = ConstantSRSadditions::Instance();
    additions->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/game_function_weights.%s.d", homePath,srsValue.c_str());
    ConstantGameFunction *gameFunction = ConstantGameFunction::Instance();
    gameFunction->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/waverage_standard_deviations.d", homePath);
    ConstantStandardDeviations *stdDevs = ConstantStandardDeviations::Instance();
    stdDevs->initialize(path);

    //read in the averages and waverages for teamA
    Team *teamA = new Team(teamNameA);
    sprintf(path, "%s/cpp/NCAA_C/teams/%i/teams.%s.averages.d",
            homePath,teamA->getYear(),boost::replace_all_copy(teamNameA," ","_").c_str());
    teamA->addAverages(path);
    sprintf(path, "%s/cpp/NCAA_C/teams/%i/teams.%s.waverages.d",
            homePath,teamA->getYear(),boost::replace_all_copy(teamNameA," ","_").c_str());
    teamA->addWAverages(path);

    //read in the averages and waverages for teamA
    Team *teamB = new Team(teamNameB);
    sprintf(path, "%s/cpp/NCAA_C/teams/%i/teams.%s.averages.d",
            homePath,teamB->getYear(),boost::replace_all_copy(teamNameB," ","_").c_str());
    teamB->addAverages(path);
    sprintf(path, "%s/cpp/NCAA_C/teams/%i/teams.%s.waverages.d",
            homePath,teamB->getYear(),boost::replace_all_copy(teamNameB," ","_").c_str());
    teamB->addWAverages(path);

    TFile *histsFile;
    std::unordered_map<int, TH1F*> probs_by_year, probs_err_by_year;
    if (useHistogramsFile){
        sprintf(path, "%s/cpp/NCAA_C/rootFiles/gameFunctionHistograms.%s.root", homePath, srsValue.c_str());
        histsFile = new TFile(path);

        for (int y = 2007; y <= 2016; y++){
            sprintf(path, "probs_by_year_%d", y);
            probs_by_year.emplace(y, (TH1F*)gROOT->FindObject(path));
            sprintf(path, "probs_err_by_year_%d", y);
            probs_err_by_year.emplace(y, (TH1F*)gROOT->FindObject(path));
        }
    }

    //create the output ROOT file
    sprintf(path, "%s/cpp/NCAA_C/rootFiles/%s",homePath,outFileName.c_str());
    TFile outFile(path,"recreate");

    //create and fill the containers for the averages and waverages
    std::unordered_map<std::string, TH1F*> avgsA_per_time, wavgsA_per_time;
    std::unordered_map<std::string, TH1F*> avgsB_per_time, wavgsB_per_time;
    std::unordered_map<std::string, TLine*> lines;
    std::string stats[] = {"oor.p", "oefg.p", "oto.p", "oftmr.p", "srs",
                           "dor.p", "defg.p", "dto.p", "dftmr.p", "rpi"};
    for (std::string &s : stats){
        sprintf(histName, "avgA_per_time_%s",(boost::replace_all_copy(s,".","").c_str()));
        avgsA_per_time.emplace(s, new TH1F(histName,s.c_str(),150,0,150));
        sprintf(histName, "wavgA_per_time_%s",(boost::replace_all_copy(s,".","").c_str()));
        wavgsA_per_time.emplace(s, new TH1F(histName,s.c_str(),150,0,150));

        sprintf(histName, "avgB_per_time_%s",(boost::replace_all_copy(s,".","").c_str()));
        avgsB_per_time.emplace(s, new TH1F(histName,s.c_str(),150,0,150));
        sprintf(histName, "wavgB_per_time_%s",(boost::replace_all_copy(s,".","").c_str()));
        wavgsB_per_time.emplace(s, new TH1F(histName,s.c_str(),150,0,150));
    }
    if (useHistogramsFile) {
        avgsA_per_time.emplace("gameScore", new TH1F("avgA_per_time_gameScore", "gameScore", 150,0,150));
        wavgsA_per_time.emplace("gameScore", new TH1F("wavgA_per_time_gameScore", "gameScore", 150,0,150));

        avgsB_per_time.emplace("gameScore", new TH1F("avgB_per_time_gameScore", "gameScore", 150,0,150));
        wavgsB_per_time.emplace("gameScore", new TH1F("wavgB_per_time_gameScore", "gameScore", 150,0,150));
    }

    //create the lines that hold the league averages for each stat
    lines.emplace("oor.p", new TLine(0,averages->get(year,"oor.p"),150,
                                     averages->get(year,"oor.p")));
    lines["oor.p"]->Write("oorpLine");
    lines.emplace("dor.p", new TLine(0,averages->get(year,"oor.p"),150,
                                     averages->get(year,"oor.p")));
    lines["dor.p"]->Write("dorpLine");
    lines.emplace("oefg.p", new TLine(0,averages->get(year,"oefg.p"),150,
                                      averages->get(year,"oefg.p")));
    lines["oefg.p"]->Write("oefgpLine");
    lines.emplace("defg.p", new TLine(0,averages->get(year,"oefg.p"),150,
                                      averages->get(year,"oefg.p")));
    lines["defg.p"]->Write("defgpLine");
    lines.emplace("oftmr.p", new TLine(0,averages->get(year,"oftmr.p"),150,
                                       averages->get(year,"oftmr.p")));
    lines["oftmr.p"]->Write("oftmrpLine");
    lines.emplace("dftmr.p", new TLine(0,averages->get(year,"oftmr.p"),150,
                                       averages->get(year,"oftmr.p")));
    lines["dftmr.p"]->Write("dftmrpLine");
    lines.emplace("oto.p", new TLine(0,averages->get(year,"oto.p"),150,
                                     averages->get(year,"oto.p")));
    lines["oto.p"]->Write("otopLine");
    lines.emplace("dto.p", new TLine(0,averages->get(year,"oto.p"),150,
                                     averages->get(year,"oto.p")));
    lines["dto.p"]->Write("dtopLine");
    lines.emplace("srs", new TLine(0,0,150,0));
    lines["srs"]->Write("srsLine");
    lines.emplace("rpi", new TLine(0,0.5,150,0.5));
    lines["rpi"]->Write("rpiLine");
    lines.emplace("gameScoreLine", new TLine(0,0.5,150,0.5));
    lines["gameScoreLine"]->Write("gameScoreLine");

    //bring the averages here locally
    std::unordered_map<std::string, TeamAverage *> averageHashA = teamA->getAveragesByDate();
    std::unordered_map<std::string, TeamAverage *> averageHashB = teamB->getAveragesByDate();

    //some placeholders
    TeamAverage *last_average;
    TeamWAverage *waverage, *last_waverage;

    //some calendar tools
    boost::gregorian::date_duration dateDuration(1);
    boost::gregorian::days indDuration;

    //cycle through the averages hash for teamA
    for (auto &average : averageHashA){
        last_average = teamA->AverageOnDate(average.second->getDate() - dateDuration);
        if (!last_average) continue; //skip the first date
        waverage = teamA->WAverageOnDate(average.second->getDate());
        last_waverage = teamA->WAverageOnDate(average.second->getDate() - dateDuration);

        int last_num_games = last_average->getNum_games();
        indDuration = average.second->getDate() - seasonInfo->get(teamA->getYear(),"season start");
        int ind = (int)indDuration.days();

        //cycle through the strings
        for (std::string &s : stats) {
            if (boost::contains(s, ".p") == 1){
                avgsA_per_time[s]->SetBinContent(ind,average.second->getValue(s));
                wavgsA_per_time[s]->SetBinContent(ind,waverage->getValue(s));
            }
            else if (boost::contains(s, ".m") == 1){
                //makes have to take into account the fact that I store the total number of makes
                //(or total number of weighted attempts) for each day, not the average.
                avgsA_per_time[s]->SetBinContent(ind,average.second->getValue(s) / (double)average.second->getNum_games());
                wavgsA_per_time[s]->SetBinContent(ind,waverage->getValue(s) / (double)waverage->getNum_games());
            }
            else if (s == "srs") {
                avgsA_per_time[s]->SetBinContent(ind,waverage->getOrigSRS());
                wavgsA_per_time[s]->SetBinContent(ind,waverage->getSrs());
            }
            else if (s == "rpi") {
                wavgsA_per_time[s]->SetBinContent(ind,waverage->getRpi());
            }
        }
        if (useHistogramsFile) {
            double gameScore = gameFunction->predictGame(waverage, teamA->getYear(), "neutral", "neutral");
            double gameScorePct, gameScorePct_err;
            if (gameScore >= 3) {
                gameScorePct = 1;
                gameScorePct_err = 1;
            }
            else if (gameScore <= -3) {
                gameScorePct = 0;
                gameScorePct_err = 1;
            }
            else {
                gameScorePct = probs_by_year[teamA->getYear()]->GetBinContent((gameScore + 3) * 1600 / 6.0);
                gameScorePct_err = probs_err_by_year[teamA->getYear()]->
                        GetBinContent((gameScore + 3) * 1600 / 6.0);
            }
            avgsA_per_time["gameScore"]->SetBinContent(ind, gameScorePct);
            avgsA_per_time["gameScore"]->SetBinError(ind, gameScorePct_err);
            wavgsA_per_time["gameScore"]->SetBinContent(ind, gameScorePct);
            wavgsA_per_time["gameScore"]->SetBinError(ind, gameScorePct_err);
        }
    }

    //cycle through the averages hash for teamB
    for (auto &average : averageHashB){
        last_average = teamB->AverageOnDate(average.second->getDate() - dateDuration);
        if (!last_average) continue; //skip the first date
        waverage = teamB->WAverageOnDate(average.second->getDate());
        last_waverage = teamB->WAverageOnDate(average.second->getDate() - dateDuration);

        int last_num_games = last_average->getNum_games();
        indDuration = average.second->getDate() - seasonInfo->get(teamB->getYear(),"season start");
        int ind = (int)indDuration.days();

        //cycle through the strings
        for (std::string &s : stats) {
            if (boost::contains(s, ".p") == 1){
                avgsB_per_time[s]->SetBinContent(ind,average.second->getValue(s));
                wavgsB_per_time[s]->SetBinContent(ind,waverage->getValue(s));
            }
            else if (boost::contains(s, ".m") == 1){
                //makes have to take into account the fact that I store the total number of makes
                //(or total number of weighted attempts) for each day, not the average.
                avgsB_per_time[s]->SetBinContent(ind,average.second->getValue(s) / (double)average.second->getNum_games());
                wavgsB_per_time[s]->SetBinContent(ind,waverage->getValue(s) / (double)waverage->getNum_games());
            }
            else if (s == "srs") {
                avgsB_per_time[s]->SetBinContent(ind,waverage->getOrigSRS());
                wavgsB_per_time[s]->SetBinContent(ind,waverage->getSrs());
            }
            else if (s == "rpi") {
                wavgsB_per_time[s]->SetBinContent(ind,waverage->getRpi());
            }
        }
        if (useHistogramsFile) {
            double gameScore = gameFunction->predictGame(waverage, teamB->getYear(), "neutral", "neutral");
            double gameScorePct, gameScorePct_err;
            if (gameScore >= 3) {
                gameScorePct = 1;
                gameScorePct_err = 1;
            }
            else if (gameScore <= -3) {
                gameScorePct = 0;
                gameScorePct_err = 1;
            }
            else {
                gameScorePct = probs_by_year[teamB->getYear()]->GetBinContent((gameScore + 3) * 1600 / 6.0);
                gameScorePct_err = probs_err_by_year[teamB->getYear()]->
                        GetBinContent((gameScore + 3) * 1600 / 6.0);
            }
            avgsB_per_time["gameScore"]->SetBinContent(ind, gameScorePct);
            avgsB_per_time["gameScore"]->SetBinError(ind, gameScorePct_err);
            wavgsB_per_time["gameScore"]->SetBinContent(ind, gameScorePct);
            wavgsB_per_time["gameScore"]->SetBinError(ind, gameScorePct_err);
        }
    }

    TeamWAverage *waA= teamA->WAverageOnDate(seasonInfo->get(teamA->getYear(),"tournament start"));
    TeamWAverage *waB= teamB->WAverageOnDate(seasonInfo->get(teamB->getYear(),"tournament start"));

    std::unordered_map<std::string, double> predictionsA = functions->predictStats(waA, waB, teamA->getYear());
    std::unordered_map<std::string, double> predictionsB = functions->predictStats(waB, waA, teamB->getYear());

    wavgsA_per_time["oor.p"]->SetBinContent(140,predictionsA["oor.p"]*waA->getValue("oor.p"));
    wavgsB_per_time["oor.p"]->SetBinContent(140,predictionsB["oor.p"]*waB->getValue("oor.p"));
    wavgsA_per_time["oefg.p"]->SetBinContent(140,predictionsA["oefg.p"]*waA->getValue("oefg.p"));
    wavgsB_per_time["oefg.p"]->SetBinContent(140,predictionsB["oefg.p"]*waB->getValue("oefg.p"));
    wavgsA_per_time["oftmr.p"]->SetBinContent(140,predictionsA["oftmr.p"]*waA->getValue("oftmr.p"));
    wavgsB_per_time["oftmr.p"]->SetBinContent(140,predictionsB["oftmr.p"]*waB->getValue("oftmr.p"));
    wavgsA_per_time["oto.p"]->SetBinContent(140,predictionsA["oto.p"]*waA->getValue("oto.p"));
    wavgsB_per_time["oto.p"]->SetBinContent(140,predictionsB["oto.p"]*waB->getValue("oto.p"));

    double gameScoreA = gameFunction->predictGame(waA,waB,teamA->getYear(),"neutral","neutral");
    double gameScoreB = gameFunction->predictGame(waB,waA,teamB->getYear(),"neutral","neutral");

    double gamePctA = probs_by_year[teamA->getYear()]->GetBinContent((gameScoreA + 3) * 1600 / 6.0);
    double gamePctB = probs_by_year[teamB->getYear()]->GetBinContent((gameScoreB + 3) * 1600 / 6.0);
    double gamePctA_err = probs_err_by_year[teamA->getYear()]->GetBinContent((gameScoreA + 3) * 1600 / 6.0);
    double gamePctB_err = probs_err_by_year[teamB->getYear()]->GetBinContent((gameScoreB + 3) * 1600 / 6.0);

    wavgsA_per_time["gameScore"]->SetBinContent(140,gamePctA);
    wavgsB_per_time["gameScore"]->SetBinContent(140,gamePctB);
    wavgsA_per_time["gameScore"]->SetBinError(140,gamePctA_err);
    wavgsB_per_time["gameScore"]->SetBinError(140,gamePctB_err);

    outFile.Write();
    outFile.Close();

    return 0;
}

void printOptions(){
    std::cout << std::endl;
    std::cout << "showTournamentMatchup Usage options" << std::endl;
    std::cout << "" << std::endl;
    std::cout << "\t-t (string) teamnameA (no default)[Required]" << std::endl;
    std::cout << "\t-T (string) teamnameB (no default)[Required]" << std::endl;
    std::cout << "\t-o (string) outFileName (no default)[Required]" << std::endl;
    std::cout << "\t-s (double) srs value to use for weights and histograms (default: \tfree\")[Optional]" << std::endl;
    std::cout << "\t-H calculate game score using histograms file (no default)[Optional]" << std::endl;
    std::cout << "" << std::endl;
    std::cout << "Ex: showTournamentMatchup -t \"2015 north carolina\" -T \"2015 harvard\" -o \"temp.root\" -H -s 0.5" << std::endl;
    std::cout << std::endl;
    std::cout << "This program creates histograms for each of the four factors, offense and" << std::endl;
    std::cout << "defense, over the course of the season for two teams.  It creates histograms for both" << std::endl;
    std::cout << "the regular averages as well as the weighted averages.  It also plots" << std::endl;
    std::cout << "SRS and gameScore or RPI.  Lastly it calculates the predictions for both teams'" << std::endl;
    std::cout << "four factors, and likelihood of winning.  You view the resultant ROOT histograms" << std::endl;
    std::cout << "using the showTournamentMatch.rb Ruby program.  All you have to specify is the" << std::endl;
    std::cout << "teams' names, and the name of the output file.  You can also decide to " << std::endl;
    std::cout << "plot the game score by selecting the -H option, and specifying an" << std::endl;
    std::cout << "SRS value for the weights if necessary." << std::endl;
    std::cout << std::endl;
}